tikv-client/src/main/java/com/pingcap/tikv/TwoPhaseCommitter.java

```diff
@@ -310,13 +313,45 @@ private void doPrewriteSecondaryKeysInBatchesWithRetry(
 
     // For prewrite, stop sending other requests after receiving first error.
     for (BatchKeys batchKeys : batchKeyList) {
-      doPrewriteSecondaryKeySingleBatchWithRetry(backOffer, primaryKey, batchKeys, mutations);
+      TiRegion oldRegion = batchKeys.getRegion();
+      TiRegion currentRegion = this.regionManager.getRegionById(oldRegion.getId());
+      if (oldRegion.equals(currentRegion)) {
+        doPrewriteSecondaryKeySingleBatchWithRetry(backOffer, primaryKey, batchKeys, mutations);
+      } else {
+        if (level > MAX_RETRY_LEVEL) {
+          throw new TiBatchWriteException(
+              String.format(
+                  "> max retry number %s, oldRegion=%s, currentRegion=%s",
+                  MAX_RETRY_LEVEL, oldRegion, currentRegion));
+        }
+        LOG.debug(
+            String.format(
+                "oldRegion=%s != currentRegion=%s, will refetch region info and retry",
+                oldRegion, currentRegion));
+        retryPrewriteBatch(
+            backOffer, primaryKey, batchKeys, mutations, level == -1 ? 1 : level + 1);
```

Comment:
why do we need to check -1 here?  What if we pass 0 rather than -1 into this function?

Reply:
The logic was correct, but we may prettify the code into:
```suggestion
            backOffer, primaryKey, batchKeys, mutations, level + 1);
```
and call `retryPrewriteBatch(.., 0)`

Reply:
Yes.

Reply:
fixed